STYLE-GUIDE.md

```diff
@@ -0,0 +1,285 @@
+  | CoolFloat SpecialFloat
+```
+
+## Imports
+
+### Don't qualify imports with single letters
+
+Qualified imports deserve intelligent names, like any other variable.  Some common instances are caught by `hlint`:
+
+```haskell
+import Data.Text qualified as T -- bad; hlint will catch this specific case
+import Data.Text qualified as Text -- Good!
+```
+
+### Don't use implicit blanket imports
+
+Import items explicitly:
+
+```haskell
+import Control.Effect.Diagnostics -- bad, which items do you need from this module?
+import Control.Effect.Diagnostics (fatal, context)  -- Good
+```
+
+There are notable exceptions, but they are few and far between.  The `Path` module is a common exception, but there's no good justification for this.
+
+## Strings
+
+### Don't use the bare conversion routines
+
+For converting between `Bytestring`/`Text`/`String`, including the lazy forms of `Bytestring` and `Text`, use the internal `Data.String.Conversion`
+module.  We export as many string conversion routines as we've found use for, plus more.  Use these instead of `decodeUtf8`/`encodeUTF8`/`Text.pack`
+/`Text.unpack`/`toFilePath`.  `hlint` will catch some of these, but not all.
+
+## Parsing
+
+We use `megaparsec` for text parsing, with a few common idioms.
+
+### Define a `Parser` type alias
+
+Usually, you should add `type Parser = Parsec Void Text` to your parsing module, so that your parser functions can be of type `Parser a`, rather than
+the long-form `Parsec` version, or worse, the fully-verbose `ParsecT` version.
+
+### Use `sc`, `scn`, `symbol`, and `lexeme`
+
+When writing a non-trivial text parser, you should create the following helper functions:
+
+- `sc`/`scn` - **S**pace **C**onsumer (or **S**pace **C**onsumer with **N**ewlines)
+  - Use [`Lexer.space`](https://hackage.haskell.org/package/megaparsec-9.1.0/docs/Text-Megaparsec-Char-Lexer.html#v:space) to create your whitespace consumer.
+  - READ THE DOCS FOR THAT FUNCTION IF YOU HAVEN'T YET!  There's a lot of useful info there, and you'll need to know it.
+  - You don't always need to create both `sc` and `scn`, but you'll almost always need at least one.
+- `symbol` - Parser for verbatim text strings.
+  - Use [`Lexer.symbol`](https://hackage.haskell.org/package/megaparsec-9.1.0/docs/Text-Megaparsec-Char-Lexer.html#v:symbol) to create this helper.
+  - Use your `sc` or `scn` function for the whitespace consumer.
+  - You can use `Lexer.symbol'` if your text is case-insensitive.
+- `lexeme` - Parser for any basic unit of the language.  While `symbol` is for verbatim text parsing, `lexeme` is used with any parser that should consume space
+  after finishing.  For example, `Lexer.symbol` is implmented via `Lexer.lexeme`.
+  - Use [`Lexer.lexeme`](https://hackage.haskell.org/package/megaparsec-9.1.0/docs/Text-Megaparsec-Char-Lexer.html#v:lexeme) to define this helper.
+  - Use your `sc` or `scn` function for the whitespace consumer.
+
+### Every parser must consume something to succeed
```

Comment:
Another tip is that parsers should generally also force `eof` at the end, to make sure they've consumed all input.

Reply:
Actually, I'm thinking we leave this section and that rule out of this guide.  It has nothing to do with style, and is more about "How to write good spectrometer code".  Maybe we should add this and other relevant stuff to `HACKING`.

Reply:
Agreed on moving it out of the style guide and into a separate "parsing best practices" doc or similar

Reply:
Moved.